### PR TITLE
fix(combobox): grouped combobox scroll

### PR DESCRIPTION
**Pull request by andipants12**
## Description
Issue reported: [Issue 2308](https://github.com/twilio-labs/paste/issues/2308)

**tl;dr** Adds a conditional guard to a side effect hook that syncs virtual scroll position with current highlighted index when the combobox has grouped items.

### Summary of changes
Grouped comboboxes are not yet virtualized because we cannot yet do so accessibly. The main component ([Combobox.tsx](https://github.com/twilio-labs/paste/pull/2335/files#diff-22b2d509777ec919245d83593056eb5d0c5ac23808f3b543b288425b985cb360L106)) was running a side effect (`useEffect`) that syncs the virtualized list position according to the highlighted index. This caused increasingly erratic scrolling behavior (with continued scrolling) because the scrolling function is not "calibrated" correctly for the non-virtualized grouped list.

To resolve this we:
- check if the combobox is grouped by determining if a key was provided by which to group items (`groupItemsBy` prop)
- If grouped, the effect logic is skipped; if not, it runs
- added unit tests for grouped, controlled, and basic comboboxes to confirm that the scroll function runs or not as is appropriate

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: packages/paste-codemods/tools/.cache/mappings.json [hunk 1]
```diff
@@ -20,6 +20,7 @@
   "Card": "@twilio-paste/core/card",
   "ChatBubble": "@twilio-paste/core/chat-log",
   "ChatMessage": "@twilio-paste/core/chat-log",
+  "MessageVariantContext": "@twilio-paste/core/chat-log",
```
Comment (andipants12) [inline packages/paste-codemods/tools/.cache/mappings.json:23]: This is odd, I didn't modify anything in codemods, but I saw this when I went to commit changes. 
I think this component was recently added, but shouldn't this diff be in a different pull request? 🤔

File: packages/paste-core/components/combobox/__tests__/index.spec.tsx [hunk 2]
```diff
@@ -277,6 +317,22 @@ describe('Combobox', () => {
...
     });
+
+    it('should not run react virtual\'s "scroll to" function when "groupItemsBy" is defined as a string', () => {
+      // bc grouped comboboxes are not yet virtualized
+      render(<GroupedMockCombobox />, {wrapper: ThemeWrapper});
+
+      const targetIndex = 1;
+      const target = groupedItems[targetIndex];
+
+      userEvent.hover(screen.getByText(target.label));
```
Comment (andipants12) [inline packages/paste-core/components/combobox/__tests__/index.spec.tsx:328]: Hover to trigger state change `highlightedIndex`, which controls the effect.

File: packages/paste-core/components/combobox/__tests__/index.spec.tsx [hunk 3]
```diff
@@ -277,6 +317,22 @@ describe('Combobox', () => {
...
+
+      expect(screen.getByRole('textbox').getAttribute('aria-activedescendant')).toMatch(
```
Comment (andipants12) [inline packages/paste-core/components/combobox/__tests__/index.spec.tsx:330]: Indirect confirmation that the correct item is active (which indirectly means it is highlighted).

File: packages/paste-core/components/combobox/__tests__/index.spec.tsx [hunk 4]
```diff
@@ -277,6 +317,22 @@ describe('Combobox', () => {
...
+        /downshift-([1-9]\d\d|[1-9]\d|\d)-item-1/g
```
Comment (andipants12) [inline packages/paste-core/components/combobox/__tests__/index.spec.tsx:331]: Just want to make sure we have the correct item index, don't really care about the uuid prefix from downshift.

File: packages/paste-core/components/combobox/__tests__/index.spec.tsx [hunk 5]
```diff
@@ -277,6 +317,22 @@ describe('Combobox', () => {
...
+      );
+
+      expect(mockScrollToIndex).not.toHaveBeenCalled();
```
Comment (andipants12) [inline packages/paste-core/components/combobox/__tests__/index.spec.tsx:334]: ✅ I verified that this expect statement fails when that guard is removed (meaning the scroll function is in fact called).

File: packages/paste-core/components/combobox/__tests__/index.spec.tsx (hunk 2 above)
Comment (shleewhite) [reply packages/paste-core/components/combobox/__tests__/index.spec.tsx:328]: Oh, this answers one of my questions above - so even if the list doesnt actually scroll it still triggers the state change?

**Comment by shleewhite** [comment]
Looks way better! I tried it out in a CodeSandbox: https://codesandbox.io/s/twilio-paste-nextjs-template-forked-9sze64?file=/pages/index.tsx

The only thing I noticed is that if an option is partially in view, it doesn't snap to be in view when it is highlighted... but I guess that is expected now because virtual is what handles that.

File: .changeset/loud-bags-jog.md [hunk 6]
```diff
@@ -0,0 +1,6 @@
+---
+'@twilio-paste/combobox': minor
+'@twilio-paste/core': minor
+---
+
+[Fix] Remediate erratic scrolling behavior for grouped combobox. Scroll function for the virtualization library was being called even though grouped comboboxes are not virtualized. The jumping behavior on scroll was observed because the indices in the virtual list and the actual list did not match. Added guard to prevent the scroll function from the virtualization library from being called.
```
Comment (SiTaggart) [inline .changeset/loud-bags-jog.md:6]: suggestion: we tend to do the component name in the square brackets so that we can identify the package in the Core bundle changelog

**Review by SiTaggart — APPROVED** [review]
just a minor tweak the changeset, but the change and test coverage seem pretty great to me

File: packages/paste-core/components/combobox/__tests__/index.spec.tsx [hunk 7]
```diff
@@ -208,6 +233,21 @@ describe('Combobox', () => {
...
       expect(renderedTextbox.getAttribute('required')).toEqual('');
     });
+
+    it('should call scroll function when highlighted index changes', () => {
+      render(<ComboboxMock />, {wrapper: ThemeWrapper});
+
+      const targetIndex = 1;
+      const target = items[targetIndex];
+
+      userEvent.hover(screen.getByText(target));
```
[the remark this answers is not in the document]
Comment (andipants12) [reply packages/paste-core/components/combobox/__tests__/index.spec.tsx:243]: Initial highlighted index is -1. We need to simulate a user event because the index change is coupled to the mouseover event.

These are things internal to the primitive from dropdown.

File: packages/paste-core/components/combobox/__tests__/index.spec.tsx (hunk 2 above)
Comment (andipants12) [reply packages/paste-core/components/combobox/__tests__/index.spec.tsx:328]: Yeah the highlighted index is coupled to mouseover position, not scroll position.

**Review by shleewhite — APPROVED** [review]

File: .changeset/loud-bags-jog.md (hunk 6 above)
Comment (SiTaggart) [reply .changeset/loud-bags-jog.md:6]: Bump, still need to Component name in the square brackets [ ]
